### DISCUSSION TITLE
`cast_motion` was passed an invalid transform

### DISCUSSION THREAD

**Discussion by moongdevstudio** [godot-jolt/godot-jolt#954] [Questions] [answered]
Godot: 4.3
Jolt: 0.13.0
OS: Win 11
Renderer: Forward+
GPU: 3080 FE

Describing Issue: Since updating to 0.13.0 and Godot 4.3, I'm getting  a few thousands error per minute and I can't figure out the root of it.

E 0:00:14:0955   _cast_motion: cast_motion was passed an invalid transform. A scale of (2.793100, 3.403400, 2.142000) is not supported by Godot Jolt for this shape/body. The scale will instead be treated as (2.779500, 2.779500, 2.779500).
  <Source C++>   src\spaces\jolt_physics_direct_space_state_3d.cpp:267 @ _cast_motion()
E 0:00:14:0955   _rest_info: get_rest_info was passed an invalid transform. A scale of (2.793100, 3.403400, 2.142000) is not supported by Godot Jolt for this shape/body. The scale will instead be treated as (2.779500, 2.779500, 2.779500).
  <Source C++>   src\spaces\jolt_physics_direct_space_state_3d.cpp:435 @ _rest_info()

There's no info on where this error is coming from in the log and my game is quite huge so it's hard to pin down the issue.

I've tried limiting my scene as much as possible but I still get the error

Is there any way to track down which node is causing this? I've tried looking up everywhere but couldn't find anyone with a similar issue. Thanks

**Comment by mihe** [comment, marked as answer]
Since you don't seem to recognize the names of `cast_motion` and `get_rest_info`, both of which are methods available on [`PhysicsDirectSpaceState3D`](https://docs.godotengine.org/en/stable/classes/class_physicsdirectspacestate3d.html), I'm guessing that you likely have some `ShapeCast3D` node somewhere, since it uses those two methods internally.

That `ShapeCast3D` node is getting scaled incorrectly, likely by some parent node, which affects the outcome of the query itself when you use Godot Jolt.

If you're not able to address the scale of the parent node you can do this as a quick hack/workaround:

1. Attach a script to your `ShapeCast3D`
2. Override `_ready` and call `orthonormalize()` at the end of it
3. Override `_physics_process` and call `orthonormalize()` at the end of it

That will normalize the transform of your `ShapeCast3D` every tick, which discards any scaling.

Another workaround is to instead change the effective node hierarchy, by utilizing `RemoteTransform3D`:

1. Add a `RemoteTransform3D` sibling next to your `ShapeCast3D`
2. Set the local transform of that `RemoteTransform3D` to the same as your `ShapeCast3D`
3. Untick the `update_scale` property of that `RemoteTransform3D`
4. Set the `remote_path` property of that `RemoteTransform3D` to point to its `ShapeCast3D` sibling
5. Add a script to your `ShapeCast3D`
6. Override `_init` and set `top_level = true`

That will make your `ShapeCast3D` a top-level node, effectively ignoring its parent/ancestors, and then the `RemoteTransform3D` will instead transform it, but without any scaling. So if you ever want to move the `ShapeCast3D` through code you'll need to instead move the `RemoteTransform3D`.

EDIT: I should probably have mentioned that the context I have available where the error is being emitted is very limited, so unfortunately I can't report what exactly is invoking the queries.

**Reply by moongdevstudio** [reply]
Hey, first off thank you very much for this detailed answer and sorry for the lack of context since the error doesn't give me anything I couldn't find anything more to add.

I found the culprit shapecast and after making it top level I did notice the transform scaling were the exact same that the one reported in the logs...

Managed to fix it by fixing the parent themselves.

I'd like to know if you can tell me how to pin down the nodes that are potential culprit in this kind of situation, like how did you find that ShapeCast3D is the only node that use both of these functions?

Thanks and sorry again for what seem like a lack of effort post.

**Reply by mihe** [reply]
> I'd like to know if you can tell me how to pin down the nodes that are potential culprit in this kind of situation, like how did you find that ShapeCast3D is the only node that use both of these functions?

That was mostly just an educated guess on my part. [Searching](https://github.com/search?q=repo%3Agodotengine%2Fgodot+path%3A%2F%5Escene%5C%2F3d%2F+%2F%5Cbcast_motion%5Cb%2F&type=code) in the Godot codebase will show you that `cast_motion` is only used in `ShapeCast3D` and `SpringArm3D`, and I figured it was more likely to be the former.

> Thanks and sorry again for what seem like a lack of effort post.

Just to clarify, I didn't mean to imply that you should have known this. I realize not everyone use the physics queries (`PhysicsDirectSpaceState3D`) directly and instead rely on "wrapper" nodes like `ShapeCast3D`, in which case this error message is borderline nonsense.

I'll change the error message to at least mention `ShapeCast3D`, in case more people get tripped up on this.

**Comment by MichaelGame-Dev** [comment]
@mihe 
I had a question related to scaling on collision shapes.

I'm working on a 3D brick breaker game called Radiant Bricks.

Basic setup is the paddles and balls are characterbody3D (I'm going with arkanoid physics, sounds like a total pain with a rigidbody3D ball). Generally with the exception of getting bricks to slide (they kinda just move the entire length) most things are working fairly well with Jolt.

I have a powerup that will scale up the size of the ball and another that scales up the size of the paddles. After the powerups end, I scale the ball and paddle back to their original size. I may later have a negative powerup that reduces the size.

I've seen that previously it has been recommended to NOT scale collision shapes. I was curious if that was still the case with Jolt and CharacterBody3Ds and if it's discouraged, what approach should I use to do this?

Currently, I'm adjusting the scale using a tween to get a nice looking animation. Do I need to use `orthonormalize()` as described here?

I was originally getting the error mentioned in the first post, but updating to 0.13 seems to have stopped the error at least from my testing so far.

**Reply by mihe** [reply]
> I've seen that previously it has been recommended to NOT scale collision shapes. I was curious if that was still the case with Jolt and CharacterBody3Ds and if it's discouraged, what approach should I use to do this?

Scaling collision _shapes_ is no problem in either of the two engines (assuming uniform scaling). Where it gets problematic is when you scale _bodies_, which Godot Physics doesn't support, and which Godot will emit warnings about in the Scene dock.

I do however support scaling bodies as well in this extension, but I've let it be an undocumented feature, since I felt weird about encouraging people to have their scene nodes show a bunch of warnings.

Anyway, it should work fine, and you should get fairly clear errors when you try to scale something in an invalid manner, so there's not much risk in terms of something silently breaking.

> Currently, I'm adjusting the scale using a tween to get a nice looking animation. Do I need to use `orthonormalize()` as described here?

`orthonormalize()` is only for situations where you don't actually want any scaling, as it will normalize the axes of your transform, leaving you with a scaling of `(1, 1, 1)`.

> I was originally getting the error mentioned in the first post, but updating to 0.13 seems to have stopped the error at least from my testing so far.

The same error-handling is still happening, but starting with 0.13.0-stable any scaling issues as a result of minor numerical precision errors in your transform (from continuously rotating or something similar) will get "repaired" behind the scenes, and you'll only see an error about it once it gets really bad.

**Reply by MichaelGame-Dev** [reply]
Thank you @mihe
Much appreciated. I am scaling the characterbody3Ds. Is it safer to just scale the mesh and collision shape instead?

Working on finishing up the the functionality pass for the moment and just working on ironing out some quirks, so just wanted to see if this was going to be an issue later in development. Thanks for godot-jolt, it's been very helpful!

**Reply by mihe** [reply]
> Is it safer to just scale the mesh and collision shape instead?

If you ever want to switch back over to Godot Physics (if only just to check something quickly) you'll want to do this instead, since Godot Physics will strip the scaling of any bodies internally, and thus differ in behavior compared to this extension.
